# Preprocessing: Analyze & Clean Data

## Import Required Libraries

In [ ]:
import pandas as pd
import numpy as np

## Access Dataset

In [ ]:
path = "/Users/dillonquick/Documents/GitHub/course-project-team-11-pedro-dillon/src/data/raw/loan_data.csv"
df = pd.read_csv(path)
display(df.head())
df.shape


,person_age,person_gender,person_education,person_income,person_emp_exp,person_home_ownership,loan_amnt,loan_intent,loan_int_rate,loan_percent_income,cb_person_cred_hist_length,credit_score,previous_loan_defaults_on_file,loan_status
0,22.0,female,Master,71948.0,0,RENT,35000.0,PERSONAL,16.02,0.49,3.0,561,No,1
1,21.0,female,High School,12282.0,0,OWN,1000.0,EDUCATION,11.14,0.08,2.0,504,Yes,0
2,25.0,female,High School,12438.0,3,MORTGAGE,5500.0,MEDICAL,12.87,0.44,3.0,635,No,1
3,23.0,female,Bachelor,79753.0,0,RENT,35000.0,MEDICAL,15.23,0.44,2.0,675,No,1
4,24.0,male,Master,66135.0,1,RENT,35000.0,MEDICAL,14.27,0.53,4.0,586,No,1


(45000, 14)

## Reading The Features

Looking at the data 2 main things stand out, person_gender & loan_int_rate. 
### Gender

Banks don't use gender as a metric to determine if someone could get a loan. Could it improve the performance of the model? Maybe but we should remove it because banks don't use that as criteria. We could just be seeing a pattern that adds noise.

### Interest Rate

If we include interest rate it would be too easy for the model to figure out who gets qualified and who doesnt. We should investigate this because banks normally set this after approval. However, In many datasets the interest rate is based on risk so it may still be valid.

### Target Variable

Our target variable will be loan_status. That tells us if someone got approved or not (0 = not approved, 1 = approved)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45000 entries, 0 to 44999
Data columns (total 14 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   person_age                      45000 non-null  float64
 1   person_gender                   45000 non-null  object 
 2   person_education                45000 non-null  object 
 3   person_income                   45000 non-null  float64
 4   person_emp_exp                  45000 non-null  int64  
 5   person_home_ownership           45000 non-null  object 
 6   loan_amnt                       45000 non-null  float64
 7   loan_intent                     45000 non-null  object 
 8   loan_int_rate                   45000 non-null  float64
 9   loan_percent_income             45000 non-null  float64
 10  cb_person_cred_hist_length      45000 non-null  float64
 11  credit_score                    45000 non-null  int64  
 12  previous_loan_defaults_on_file  

Looks like there are no Null Values to remove but we should still check if there are non valid values like if someone is not male or female

In [ ]:
display(df.describe())

,person_age,person_income,person_emp_exp,loan_amnt,loan_int_rate,loan_percent_income,cb_person_cred_hist_length,credit_score,loan_status
count,45000.000000,4.500000e+04,45000.000000,45000.000000,45000.000000,45000.000000,45000.000000,45000.000000,45000.000000
mean,27.764178,8.031905e+04,5.410333,9583.157556,11.006606,0.139725,5.867489,632.608756,0.222222
std,6.045108,8.042250e+04,6.063532,6314.886691,2.978808,0.087212,3.879702,50.435865,0.415744
min,20.000000,8.000000e+03,0.000000,500.000000,5.420000,0.000000,2.000000,390.000000,0.000000
25%,24.000000,4.720400e+04,1.000000,5000.000000,8.590000,0.070000,3.000000,601.000000,0.000000
50%,26.000000,6.704800e+04,4.000000,8000.000000,11.010000,0.120000,4.000000,640.000000,0.000000
75%,30.000000,9.578925e+04,8.000000,12237.250000,12.990000,0.190000,8.000000,670.000000,0.000000
max,144.000000,7.200766e+06,125.000000,35000.000000,20.000000,0.660000,30.000000,850.000000,1.000000


# Reading the df description

Looking at the description we can see that there is someone who is 144 years old. That is highly unlikely so we should clean that up by caping those values. We can set all the ages over 90 to 90

In [ ]:
# get ages and credit history length for people over 90
over_90_data = df.loc[df['person_age'] >= 90, ['person_age', 'cb_person_cred_hist_length']].values
over_90_data

array([[144.,   3.],
       [144.,   2.],
       [123.,   3.],
       [123.,   4.],
       [144.,  25.],
       [ 94.,  27.],
       [116.,  24.],
       [109.,  22.]])

Based off of these results we can see there are only 8 entries above the age of 90 and if we cap the age at 90 the credit history length wouldn't become contradictory. Therefore, we can proceed and cap the age at 90.

In [ ]:
df_capped = df.copy()
df_capped['person_age'] = np.where(df_capped['person_age'] > 84, 84, df_capped['person_age'])
df_capped.describe()

,person_age,person_income,person_emp_exp,loan_amnt,loan_int_rate,loan_percent_income,cb_person_cred_hist_length,credit_score,loan_status
count,45000.000000,4.500000e+04,45000.000000,45000.000000,45000.000000,45000.000000,45000.000000,45000.000000,45000.000000
mean,27.756956,8.031905e+04,5.410333,9583.157556,11.006606,0.139725,5.867489,632.608756,0.222222
std,5.948487,8.042250e+04,6.063532,6314.886691,2.978808,0.087212,3.879702,50.435865,0.415744
min,20.000000,8.000000e+03,0.000000,500.000000,5.420000,0.000000,2.000000,390.000000,0.000000
25%,24.000000,4.720400e+04,1.000000,5000.000000,8.590000,0.070000,3.000000,601.000000,0.000000
50%,26.000000,6.704800e+04,4.000000,8000.000000,11.010000,0.120000,4.000000,640.000000,0.000000
75%,30.000000,9.578925e+04,8.000000,12237.250000,12.990000,0.190000,8.000000,670.000000,0.000000
max,84.000000,7.200766e+06,125.000000,35000.000000,20.000000,0.660000,30.000000,850.000000,1.000000


In [ ]:
df_cleaned = df.copy()

df_cleaned['person_age'] = df["person_age"].clip(upper = 90)
print(df_cleaned['person_age'].max())

#removes person gender from the df as the cleaned product
df_cleaned = df_cleaned.drop(columns=['person_gender'])
display(df_cleaned.columns)


x = df_cleaned.drop(columns=['loan_status']) #all columns but loan status (training data for the model)
y = df_cleaned['loan_status'] #only loan status column (answesrs for if they are appoved or not)


display(x.head())
display(y.head())

90.0


Index(['person_age', 'person_education', 'person_income', 'person_emp_exp',
       'person_home_ownership', 'loan_amnt', 'loan_intent', 'loan_int_rate',
       'loan_percent_income', 'cb_person_cred_hist_length', 'credit_score',
       'previous_loan_defaults_on_file', 'loan_status'],
      dtype='object')

,person_age,person_education,person_income,person_emp_exp,person_home_ownership,loan_amnt,loan_intent,loan_int_rate,loan_percent_income,cb_person_cred_hist_length,credit_score,previous_loan_defaults_on_file
0,22.0,Master,71948.0,0,RENT,35000.0,PERSONAL,16.02,0.49,3.0,561,No
1,21.0,High School,12282.0,0,OWN,1000.0,EDUCATION,11.14,0.08,2.0,504,Yes
2,25.0,High School,12438.0,3,MORTGAGE,5500.0,MEDICAL,12.87,0.44,3.0,635,No
3,23.0,Bachelor,79753.0,0,RENT,35000.0,MEDICAL,15.23,0.44,2.0,675,No
4,24.0,Master,66135.0,1,RENT,35000.0,MEDICAL,14.27,0.53,4.0,586,No


0    1
1    0
2    1
3    1
4    1
Name: loan_status, dtype: int64

In [ ]:
df.head(20)

,person_age,person_gender,person_education,person_income,person_emp_exp,person_home_ownership,loan_amnt,loan_intent,loan_int_rate,loan_percent_income,cb_person_cred_hist_length,credit_score,previous_loan_defaults_on_file,loan_status
0,22.0,female,Master,71948.0,0,RENT,35000.0,PERSONAL,16.02,0.49,3.0,561,No,1
1,21.0,female,High School,12282.0,0,OWN,1000.0,EDUCATION,11.14,0.08,2.0,504,Yes,0
2,25.0,female,High School,12438.0,3,MORTGAGE,5500.0,MEDICAL,12.87,0.44,3.0,635,No,1
3,23.0,female,Bachelor,79753.0,0,RENT,35000.0,MEDICAL,15.23,0.44,2.0,675,No,1
4,24.0,male,Master,66135.0,1,RENT,35000.0,MEDICAL,14.27,0.53,4.0,586,No,1
5,21.0,female,High School,12951.0,0,OWN,2500.0,VENTURE,7.14,0.19,2.0,532,No,1
6,26.0,female,Bachelor,93471.0,1,RENT,35000.0,EDUCATION,12.42,0.37,3.0,701,No,1
7,24.0,female,High School,95550.0,5,RENT,35000.0,MEDICAL,11.11,0.37,4.0,585,No,1
8,24.0,female,Associate,100684.0,3,RENT,35000.0,PERSONAL,8.90,0.35,2.0,544,No,1
9,21.0,female,High School,12739.0,0,OWN,1600.0,VENTURE,14.74,0.13,3.0,640,No,1


In [ ]:
# Get unique entries for the specified columns
print("Finding Unique Values for Categorical Columns:")
unique_education = df['person_education'].unique()
unique_home_ownership = df['person_home_ownership'].unique()
unique_loan_intent = df['loan_intent'].unique()

print("Unique person_education:", unique_education)
print("Unique person_home_ownership:", unique_home_ownership)
print("Unique loan_intent:", unique_loan_intent)

# Find rows with any negative values in numerical columns
print("\n\nChecking for negative values in numerical columns:")
numerical_cols = ['person_age', 'person_income', 'person_emp_exp', 'loan_amnt', 'loan_int_rate', 'loan_percent_income', 'cb_person_cred_hist_length', 'credit_score', 'loan_status']
negative_rows = df[(df[numerical_cols] < 0).any(axis=1)]
print(negative_rows.shape[0], "rows with negative values found.")

#get some statistics on the "other" home ownership group to determine if we should drop it or not
print("\n\nStatistics for 'OTHER' home ownership group:")
display(df["person_home_ownership"].value_counts())
other_home_ownership = df[df['person_home_ownership'] == 'OTHER'][['person_age', 'person_income', 'loan_intent']]
display(other_home_ownership)
other_home_ownership.describe()

Finding Unique Values for Categorical Columns:
Unique person_education: ['Master' 'High School' 'Bachelor' 'Associate' 'Doctorate']
Unique person_home_ownership: ['RENT' 'OWN' 'MORTGAGE' 'OTHER']
Unique loan_intent: ['PERSONAL' 'EDUCATION' 'MEDICAL' 'VENTURE' 'HOMEIMPROVEMENT'
 'DEBTCONSOLIDATION']


Checking for negative values in numerical columns:
0 rows with negative values found.


Statistics for 'OTHER' home ownership group:


person_home_ownership
RENT        23443
MORTGAGE    18489
OWN          2951
OTHER         117
Name: count, dtype: int64

,person_age,person_income,loan_intent
29,21.0,14988.0,EDUCATION
566,26.0,24183.0,PERSONAL
715,24.0,25826.0,PERSONAL
880,21.0,28319.0,EDUCATION
1308,22.0,31844.0,DEBTCONSOLIDATION
...,...,...,...
43155,26.0,71556.0,MEDICAL
44425,25.0,71842.0,MEDICAL
44430,28.0,88047.0,MEDICAL
44498,24.0,71536.0,MEDICAL


,person_age,person_income
count,117.000000,117.000000
mean,26.897436,91480.068376
std,6.363127,94488.588654
min,21.000000,14988.000000
25%,23.000000,51596.000000
50%,24.000000,72093.000000
75%,28.000000,96409.000000
max,48.000000,721514.000000


Looking at the categorical columns we can see that all of them are valid and don't need to be dropped. We also checked if there are any negative values in the numerical columns which there isn't so we're good on that front. Now we just need to map the categroical columns.

# Mapping Categorical Variables

In [ ]:
education_map = {
    "High School": 1,
    "Associate": 2,
    "Bachelor": 3,
    "Master": 4,
    "Doctorate": 5,
}

if "person_education" in df_cleaned.columns:
    df_cleaned["person_education"] = df_cleaned["person_education"].map(education_map)

nominal_cols = ["person_home_ownership", "loan_intent"]


if nominal_cols:
    df_cleaned = pd.get_dummies(df_cleaned, columns=nominal_cols, drop_first=True, dtype=int)

display(df.head())
display(df_cleaned.head())

,person_age,person_gender,person_education,person_income,person_emp_exp,person_home_ownership,loan_amnt,loan_intent,loan_int_rate,loan_percent_income,cb_person_cred_hist_length,credit_score,previous_loan_defaults_on_file,loan_status
0,22.0,female,Master,71948.0,0,RENT,35000.0,PERSONAL,16.02,0.49,3.0,561,No,1
1,21.0,female,High School,12282.0,0,OWN,1000.0,EDUCATION,11.14,0.08,2.0,504,Yes,0
2,25.0,female,High School,12438.0,3,MORTGAGE,5500.0,MEDICAL,12.87,0.44,3.0,635,No,1
3,23.0,female,Bachelor,79753.0,0,RENT,35000.0,MEDICAL,15.23,0.44,2.0,675,No,1
4,24.0,male,Master,66135.0,1,RENT,35000.0,MEDICAL,14.27,0.53,4.0,586,No,1


,person_age,person_education,person_income,person_emp_exp,loan_amnt,loan_int_rate,loan_percent_income,cb_person_cred_hist_length,credit_score,previous_loan_defaults_on_file,loan_status,person_home_ownership_OTHER,person_home_ownership_OWN,person_home_ownership_RENT,loan_intent_EDUCATION,loan_intent_HOMEIMPROVEMENT,loan_intent_MEDICAL,loan_intent_PERSONAL,loan_intent_VENTURE
0,22.0,4,71948.0,0,35000.0,16.02,0.49,3.0,561,No,1,0,0,1,0,0,0,1,0
1,21.0,1,12282.0,0,1000.0,11.14,0.08,2.0,504,Yes,0,0,1,0,1,0,0,0,0
2,25.0,1,12438.0,3,5500.0,12.87,0.44,3.0,635,No,1,0,0,0,0,0,1,0,0
3,23.0,3,79753.0,0,35000.0,15.23,0.44,2.0,675,No,1,0,0,1,0,0,1,0,0
4,24.0,4,66135.0,1,35000.0,14.27,0.53,4.0,586,No,1,0,0,1,0,0,1,0,0
